### PR TITLE
ETS Templating

### PR DESCRIPTION
**Pull request by nllong**
@YanfeiNREL -- is this branch the branch to review for the ETS template? Also, is the `Yanfei` branch needed anymore?

If this is the right branch, then can you fill out this PR info?

* Any background on PR
* What does this PR accomplish
* How to test this PR
* What tickets does this address
* Screenshots (if applicable)

### CODE REVIEW

**Comment by nllong** [comment]
@YanfeiNREL -- can you doublecheck why the CI is failing on this PR?

**Comment by YanfeiNREL** [comment]
Pull Request Details for More Information:

(1) "template_ets" is the branch i made the templates for ETS models.

(2) "Yanfei" branch, needs to be removed, as it is a stale branch for earlier templating activities.

(3)  Background/accomplishments of the "template_ets" branch: 

      -- (3.1) I updated the Modelica folder, by adding the ETS modelica model: geojson-modelica-translator/geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/EnergyTransferStations/*. Those ETS models, examples are all from **Issue-1442** branch from **Kitty**.
                 Please see the figure below:                 
screen002_ets_template

      -- (3.2) Creating templates by changing the key parameters of ETS, through Jinja2.

      -- (3.3) ETS template is located at: geojson-modelica-translator/geojson_modelica_translator/model_connectors/ets_template.py.

      -- (3.4) The unit test of ETS-template is located:  geojson-modelica-translator/geojson_modelica_translator/tests/model_connectors/test_ets.py.
                 The unit tests all passed. Please see the figure below for details.
                 
screen001_ets_template
 
      -- (3.5) The generated ETS template was validated in Dymola (windows) only, but not in JModelica. I tried to install JModelica in my laptop, but i was failed. 
                 Please see ETS-templated model in Dymola:
screen003_ets_template
                
                 Please see simulation results of ETS-template in Dymola(district side supply water temperature, and district side return water temperature ):
                
screen004_ets_template
 

      -- (3.6)  I modified the git-ignore file to keep the updated ETS modelica models in the repo. Because my ets-template and tests are using them.

      -- (3.7) I used pip3(version 9.0.1) and python3 (version 3.6.9) to do the works.

File: .gitignore [hunk 1]
```diff
@@ -95,4 +95,4 @@ ENV/
 tests/output
 tests/modelica/output
 tests/model_connectors/output
-/geojson_modelica_translator/modelica/buildingslibrary/
+#/geojson_modelica_translator/modelica/buildingslibrary/
```
Comment (nllong) [inline .gitignore]: we don't want to commit this to github. The setup.py checks out the repo as needed.

File: README.rst [hunk 2]
```diff
@@ -20,11 +20,13 @@ The GeoJSON / Modelica Translator is still in early alpha-phase development and
...
 If installing this package for development then you must run the `setup.py build` command in order to install the MBL in the right location.
 
+currently you need python3 and pip3 to install/build the packages.
+
 .. code-block:: bash
 
-    pip install -r requirements.txt
-    python setup.py build
-    py.test
+    pip3 install -r requirements.txt
```
Comment (nllong) [inline README.rst]: let's keep this as `pip`. If the user have pip3, then they should `ln -s` it to pip. Python 2 is no longer supported and I don't see why we need to distinguish between pip 2 and pip 3.

File: geojson_modelica_translator/model_connectors/ets_cooling_indirect_yanfei.mo [hunk 3]
```diff
@@ -0,0 +1,355 @@
+
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_cooling_indirect_yanfei.mo]: This file shouldn't be named with _yanfei.mo. Perhaps just rename as ets_cooling_indirect.mo? or _v1.mo?

File: geojson_modelica_translator/model_connectors/ets_cooling_indirect_yanfei.mo [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+  
+  // mass flow rates
+  parameter Modelica.SIunits.MassFlowRate m1_flow_nominal(
+    final min=0,
+    start=0.666)
+    "Nominal mass flow rate of primary (district) district cooling side";
+  
+
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_cooling_indirect_yanfei.mo]: Is there an auto-format for MO files? Would be nice to have whitespace removed...

File: geojson_modelica_translator/model_connectors/ets_template.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            data = json.load(f)
+
+        ets_general = False
+        for key, value in data.items():
+            if key == 'definitions':
+                # three levels down to get the ETS signal
+                junctions = data["definitions"]["ThermalJunctionType"]["enum"]
+                if 'ETS' in junctions:
+                    ets_general = True
+                    print("ETS is there!!!")
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_template.py]: remove

File: geojson_modelica_translator/model_connectors/ets_template.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        with open(self.system_parameters_geojson, 'r') as f:
+            data = json.load(f)
+
+        ets_details = False
+        for key, value in data.items():
+            # four levels down to get the details
+            ets_details = data["definitions"]["building_def"]["properties"]["ets"]
+            print(ets_details)
+            if ets_details:
+                print("ETS details are here!!!")
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_template.py]: this doesn't do anything, remove.

File: geojson_modelica_translator/model_connectors/ets_template.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        # NL: This code doesn't appear to be used... review
+        # ets_modelica = ""
+        # if self.check_ets_from_building_modelica():
+        #     with open(self.ets_from_building_modelica) as f:
+        #         ets_modelica = f.read()
+        # else:
+        #     pass
+
+        # Here come the Jinja2 function: get_template()
+        ets_template = self.template_env.get_template('CoolingIndirect.mot')
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_template.py]: nice!

File: geojson_modelica_translator/model_connectors/ets_template.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        repl_dict[from_str] = to_str
+
+        with open(self.directory_modelica_building + cooling_indirect_filename, "w") as examplefile:
+            for f in file:
+                fx = f
+                for from_str, to_str in enumerate(repl_dict):
+                    print(f.__class__)
+                    # TODO: f.string() causes errors, check code
+                    if f.string() == from_str:
+                        fx = f.replace(from_str, to_str)
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/ets_template.py]: note that some of the tests are failing here

File: geojson_modelica_translator/model_connectors/templates/CoolingIndirect.mot [hunk 9]
```diff
@@ -0,0 +1,368 @@
+
+within Buildings.Applications.DHC.EnergyTransferStations;
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/templates/CoolingIndirect.mot]: looks good!

File: geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/CentralPlants/package.mo [hunk 10]
```diff
@@ -0,0 +1,10 @@
+within Buildings.Applications.DHC;
```
Comment (nllong) [inline geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/CentralPlants/package.mo]: I don't think we want modelica/buildingslibrary in github. Setup.py should handle the install.

File: tests/model_connectors/data/jm_ipython.sh [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+#################################################
+# Shell script that simulates JModelica using
+# a docker image of JModelica.
+#
+# The main purpose of this script is to export
+# MODELICAPATH and PYTHONPATH with their values
+# updated for the docker, and to mount the
+# required directories.
+#################################################
+set -e
```
Comment (nllong) [inline tests/model_connectors/data/jm_ipython.sh]: thanks for adding this file. Maybe this should go in a general location where it is accessble by the entire geojson translator?

File: tests/model_connectors/test_ets.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import os
+import unittest
+
+from geojson_modelica_translator.model_connectors.ets_template import ETS_Template
```
Comment (nllong) [inline tests/model_connectors/test_ets.py]: for some reason this test file can't run by itself???

File: .gitignore (hunk 1 above)
Comment (YanfeiNREL) [reply .gitignore]: okay, i revised it back.

**Comment by nllong** [comment]
@YanfeiNREL -- just checking on the status of this?

**Comment by YanfeiNREL** [comment]
My laptop was scheduled to install Yubi key on 1/31/2020, Friday, before the ASHRAE Winter Conference. After that, i was not able to login my VirtualMachine(Ubuntu18.04). I am waiting IT to figure out why. This is really frustrating. @nllong

File: README.rst (hunk 2 above)
Comment (nllong) [reply README.rst]: @YanfeiNREL   -- can you address this? Keep this as just pip and python.

File: geojson_modelica_translator/model_connectors/spawn.py [hunk 13]
```diff
@@ -56,16 +61,28 @@ def add_building(self, urbanopt_building, mapper=None):
...
-                "floor_height": urbanopt_building.feature.properties['height'] * 0.3048,  # ft -> m
-                "num_stories": urbanopt_building.feature.properties['number_of_stories_above_ground'],
-                "num_stories_below_grade": urbanopt_building.feature.properties['number_of_stories'] -
-                urbanopt_building.feature.properties['number_of_stories_above_ground'],
-                "year_built": urbanopt_building.feature.properties['year_built']
-            })
+            self.buildings.append(
+                {
+                    "area": urbanopt_building.feature.properties["floor_area"]
+                    * 0.092936,  # ft2 -> m2
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/spawn.py]: this formatting seems wrong?  We should allow 100 characters per line.

File: geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/EnergyTransferStations/CoolingIndirect.mo [hunk 14]
```diff
@@ -0,0 +1,334 @@
+within Buildings.Applications.DHC.EnergyTransferStations;
+model CoolingIndirect
```
Comment (nllong) [inline geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/EnergyTransferStations/CoolingIndirect.mo]: i'm not sure that we want to directly copy this into the project... leave for now. But I think we want to have the modelica library dynamically installed during the setup phase.

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+- repo: https://github.com/pre-commit/mirrors-autopep8
+  rev: v1.4.4
+  hooks:
+    - id: autopep8
```
Comment (nllong) [inline .pre-commit-config.yaml]: need to allow 100 characters per line. See this precommit config (https://github.com/urbanopt/modelica-builder/blob/develop/.pre-commit-config.yaml)

File: geojson_modelica_translator/model_connectors/spawn.py (hunk 13 above)
Comment (YanfeiNREL) [reply geojson_modelica_translator/model_connectors/spawn.py]: My pre-commit forced the changes. Maybe we can use the same pre-commit set-up to be consistent.

Comment (nllong) [reply geojson_modelica_translator/model_connectors/spawn.py]: yeah, that happens. I linked to another precommit file that I am using in another comment. You'll have to unfortunately go and fix these by hand though.

File: .pre-commit-config.yaml (hunk 15 above)
Comment (YanfeiNREL) [reply .pre-commit-config.yaml]: I think we might need to allow more than 100 characters per line. I see sometimes, 100 characters are not enough for a complete sentence. 120 characters will be better. I see in PyCharm, the maximum characters of each line is 121.

**Comment by YanfeiNREL** [comment]
Hey Nick, i see all checks are passed. Please merge ets_template into develop. Thanks.

**Comment by nllong** [comment]
> Hey Nick, i see all checks are passed. Please merge ets_template into develop. Thanks.

It looks like there are still over 211 files that are being added and I don't think that is right. Specifically all of the modelica library files...

File: .pre-commit-config.yaml (hunk 15 above)
Comment (nllong) [reply .pre-commit-config.yaml]: Sure, set this to 120!

File: tests/model_connectors/data/jm_ipython.sh (hunk 11 above)
Comment (nllong) [reply tests/model_connectors/data/jm_ipython.sh]: This is going to conflict with the this PR https://github.com/urbanopt/geojson-modelica-translator/pull/35. Resolve when we get there.

File: geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/EnergyTransferStations/CoolingIndirect.mo (hunk 14 above)
Comment (nllong) [reply geojson_modelica_translator/modelica/buildingslibrary/Buildings/Applications/DHC/EnergyTransferStations/CoolingIndirect.mo]: To clarify, I think it is okay to include a few of the MBL files here directly (for now), but it seems like we are including many more than needed.

**Pull request closed without merging**